src/agent/services/oracle.rs

```diff
@@ -0,0 +1,206 @@
+            network::{
+                Config,
+                Network,
+            },
+        },
+        state::oracle::{
+            Oracle,
+            PricePublishingMetadata,
+        },
+    },
+    anyhow::Result,
+    solana_account_decoder::UiAccountEncoding,
+    solana_client::{
+        nonblocking::{
+            pubsub_client::PubsubClient,
+            rpc_client::RpcClient,
+        },
+        rpc_config::{
+            RpcAccountInfoConfig,
+            RpcProgramAccountsConfig,
+        },
+        rpc_response::{
+            Response,
+            RpcKeyedAccount,
+        },
+    },
+    solana_sdk::{
+        account::Account,
+        commitment_config::CommitmentConfig,
+        pubkey::Pubkey,
+    },
+    std::{
+        collections::HashMap,
+        sync::Arc,
+    },
+    tokio::{
+        select,
+        sync::watch::Sender,
+    },
+    tokio_stream::StreamExt,
+};
+
+pub async fn oracle<S>(
+    config: Config,
+    network: Network,
+    state: Arc<S>,
+    publisher_permissions_tx: Sender<HashMap<Pubkey, HashMap<Pubkey, PricePublishingMetadata>>>,
+) where
+    S: Oracle,
+    S: Send + Sync + 'static,
+{
+    loop {
+        let _ = run(
+            config.clone(),
+            network,
+            state.clone(),
+            publisher_permissions_tx.clone(),
+        )
+        .await;
+    }
+}
+
+async fn run<S>(
+    config: Config,
+    network: Network,
+    state: Arc<S>,
+    publisher_permissions_tx: Sender<HashMap<Pubkey, HashMap<Pubkey, PricePublishingMetadata>>>,
+) -> Result<()>
+where
+    S: Oracle,
+    S: Send + Sync + 'static,
+{
+    let key_store = KeyStore::new(config.key_store.clone())?;
+    let mut poll_interval = tokio::time::interval(config.oracle.poll_interval_duration);
+
+    // Setup PubsubClient to listen for account changes on the Oracle program.
+    let client = PubsubClient::new(config.wss_url.as_str()).await?;
+    let (mut notifier, _unsub) = {
+        let program_key = key_store.program_key;
+        let commitment = config.oracle.commitment;
+        let config = RpcProgramAccountsConfig {
+            account_config: RpcAccountInfoConfig {
+                commitment: Some(CommitmentConfig { commitment }),
+                encoding: Some(UiAccountEncoding::Base64Zstd),
+                ..Default::default()
+            },
+            filters:        None,
+            with_context:   Some(true),
+        };
+        client.program_subscribe(&program_key, Some(config)).await
+    }?;
+
+    // Setup an RpcClient for manual polling.
+    let client = {
+        Arc::new(RpcClient::new_with_timeout_and_commitment(
+            config.rpc_url,
+            config.rpc_timeout,
+            CommitmentConfig {
+                commitment: config.oracle.commitment,
+            },
+        ))
+    };
+
+    loop {
+        select! {
+            update = notifier.next()      => handle_account_update(state.clone(), network, update).await,
```

Comment:
how we handle websocket failures here? i don't see any logic to handle it. i think we had it in the past.